namesys/mpns.go

```diff
@@ -0,0 +1,310 @@
+package namesys
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+	"github.com/ipfs/boxo/ipns"
+	"github.com/ipfs/boxo/path"
+	"github.com/ipfs/go-cid"
+	ds "github.com/ipfs/go-datastore"
+	dssync "github.com/ipfs/go-datastore/sync"
+	ci "github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/routing"
+	"github.com/miekg/dns"
+	madns "github.com/multiformats/go-multiaddr-dns"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// nameSys is a multi-protocol [NameSystem] that implements generic IPFS naming.
+// It uses several [Resolver]s:
+//
+//  1. IPFS routing naming: SFS-like PKI names.
+//  2. dns domains: resolves using links in DNS TXT records
+//
+// It can only publish to: 1. IPFS routing naming.
+type nameSys struct {
```

Comment:
nit:  "namesys" has always been lowercase, either do that or keep old name

```suggestion
type namesys struct {
```

```suggestion
type mpns struct {
```